### DISCUSSION TITLE
Help with getting Trace data in console

### DISCUSSION THREAD

**Discussion by bgouveia5** [jankae/LibreVNA#257] [General]
I'm writing a python program that gets data from the VNA but im having an issue with the data not being received. 

import sys
import time
from libreVNA import libreVNA

# Create the control instance
vna = libreVNA('localhost', 19542)

# Quick connection check (should print "LibreVNA-GUI")
print(vna.query("*IDN?"))

def safe_query(query, vna, timeout=None):
    try:
        response = vna.query(query, timeout=timeout)
        print(f"Query: {query}, Response: {response}")
        return response
    except Exception as e:
        print(f"Error during query '{query}': {e}")
        return None  # Allow the script to continue

def safe_cmd(cmd, vna, check=False, timeout=None):
    try:
        vna.cmd(cmd, check=check, timeout=timeout)
        print(f"Command: {cmd} executed successfully")
    except Exception as e:
        print(f"Error during command '{cmd}': {e}")

# Quick connection check (should print "LibreVNA-GUI")
idn_response = safe_query("*IDN?", vna)
dev = safe_query(":DEV:CONN?", vna)

if dev is None or "ERROR" in dev or "Not connected" in dev:
    print("Device is not connected properly. Aborting")
    sys.exit(-1)
else:
    print(f"Connected to {dev}")

# Simple trace data extraction

# Switch to VNA mode, setup the sweep parameters
print("Setting up sweep parameters...")
safe_cmd(":DEV:MODE VNA", vna)
time.sleep(0.5)
safe_cmd(":VNA:SWEEP FREQUENCY", vna)
time.sleep(0.5)
safe_cmd(":VNA:STIM:LVL -10", vna)
time.sleep(0.5)
safe_cmd(":VNA:ACQ:IFBW 10000", vna)
time.sleep(0.5)
safe_cmd(":VNA:ACQ:AVG 2", vna)
time.sleep(0.5)
safe_cmd(":VNA:ACQ:POINTS 501", vna)
time.sleep(0.5)
safe_cmd(":VNA:FREQuency:START 250000000", vna)
time.sleep(0.5)
safe_cmd(":VNA:FREQuency:STOP 350000000", vna)
time.sleep(0.5)
# Manually check status
print("Manual status check...")
status = safe_query("*ESR?", vna)
if status is None or "Not connected" in status:
    print("Device status check failed. Continuing without status check.")
time.sleep(0.5)

# Wait for the sweep to finish
print("Waiting for the sweep to finish...")
try:
    while safe_query(":VNA:ACQ:FIN?", vna) == "FALSE":
        time.sleep(0.5)
except Exception as e:
    print(f"Error during sweep acquisition: {e}")

# Grab the data of trace S11
print("Reading trace data...")
data = safe_query(":VNA:TRACE:DATA? S11", vna)
time.sleep(0.5)
if data:
    try:
        S11 = vna.parse_VNA_trace_data(data)
        time.sleep(0.5)
        for x in S11:
            print(x)       
    except Exception as e:
        print(f"Error reading trace data: {e}")
else:
    print("No data received for trace S11.")`
The output in console I'm getting
LibreVNA,LibreVNA-GUI,dummy_serial,1.5.1
Query: *IDN?, Response: LibreVNA,LibreVNA-GUI,dummy_serial,1.5.1
Query: :DEV:CONN?, Response: 205E35C43750
Connected to 205E35C43750
Setting up sweep parameters...
Command: :DEV:MODE VNA executed successfully
Command: :VNA:SWEEP FREQUENCY executed successfully
Command: :VNA:STIM:LVL -10 executed successfully
Command: :VNA:ACQ:IFBW 10000 executed successfully
Command: :VNA:ACQ:AVG 2 executed successfully
Command: :VNA:ACQ:POINTS 501 executed successfully
Command: :VNA:FREQuency:START 250000000 executed successfully
Command: :VNA:FREQuency:STOP 350000000 executed successfully
Manual status check...
Query: *ESR?, Response: 
Waiting for the sweep to finish...
Query: :VNA:ACQ:FIN?, Response: 
Reading trace data...
Query: :VNA:TRACE:DATA? S11, Response: 
No data received for trace S11.

I have the firmware on the VNA as 1.5.1. From the output everything is working but "Query: :VNA:TRACE:DATA? S11, Response: 
No data received for trace S11." I don't know how I could fix this.

**Comment by jankae** [comment]
There is an example script which does pretty much what you want to do: https://github.com/jankae/LibreVNA/blob/master/Documentation/UserManual/SCPI_Examples/retrieve_trace_data.py

Please try if that works.

If you are using version 1.5.1 be sure that you also checkout this script (and, more importantly, the LibreVNA.py) from the v1.5.1 tag. There is some work on master which changes the SCPI behavior a little bit.

**Reply by bgouveia5** [reply]
If I run the retrieve_trace_data i get this error. libreVNA.py:114 in get_status
    raise Exception("Expected numeric response from *ESR? but got "
Exception: Expected numeric response from *ESR? but got ''. 

I just got it able to work, by commenting all this code 
#print("Setting up sweep parameters...")
#safe_cmd(":DEV:MODE VNA", vna)
#time.sleep(0.5)
#safe_cmd(":VNA:SWEEP FREQUENCY", vna)
#time.sleep(0.5)
#safe_cmd(":VNA:STIM:LVL -10", vna)
#time.sleep(0.5)
#safe_cmd(":VNA:ACQ:IFBW 10000", vna)
#time.sleep(0.5)
#safe_cmd(":VNA:ACQ:AVG 2", vna)
#time.sleep(0.5)
#safe_cmd(":VNA:ACQ:POINTS 501", vna)
#time.sleep(0.5)
#safe_cmd(":VNA:FREQuency:START 250000000", vna)
#time.sleep(0.5)
#safe_cmd(":VNA:FREQuency:STOP 350000000", vna)
#time.sleep(0.5)
I think it has to deal with python talking to the VNA to fast, to the point it has trouble with receiving the data. The simple fix to this is adding another delay to the process.

**Reply by jankae** [reply]
> If I run the retrieve_trace_data i get this error. libreVNA.py:114 in get_status
raise Exception("Expected numeric response from *ESR? but got "
Exception: Expected numeric response from *ESR? but got ''.

Then you didn't use the script (and LibreVNA.py!) from the v1.5.1 tag as I told you. You can not mix the 1.5.1 GUI with the script examples from master.

> I think it has to deal with python talking to the VNA to fast, to the point it has trouble with receiving the data

I doubt that. I can run scripts without delays and do so in the automated tests. 

By the way: if you want to share code, markdown has the formatting for that. It makes the code much more readable. Like this:
```python
print("Reading trace data...")
data = safe_query(":VNA:TRACE:DATA? S11", vna)
time.sleep(0.5)
```

**Reply by bgouveia5** [reply]
I did use the LibreVNA.py script, sorry that I didn't relay that. Thank you for telling me about markdown, I don't use GitHub all that much.

**Reply by jankae** [reply]
Yes, I can see that you used the libreVNA.py from your script:
```python
from libreVNA import libreVNA
```
But I told you to use the libreVNA.py from the v1.5.1 tag if you use GUI version 1.5.1 and I can also see that you are not doing that from your error message. The `*ESR?` command is not available in 1.5.1. It is neither in the GUI, the programming guide or the libreVNA.py.

It will be available in the next release and the files (programming guide, libreVNA.py...) on master already include that change.

**But you can not use these latest files from master in combination with the 1.5.1 GUI.** There were changes in the behavior of the SCPI server, if you want the details, you can read this issue: #248

The fact that you are getting the `Expected numeric response from *ESR? but got ''` message is a clear indicator to me that you are using at least the libreVNA.py from master. This does not work and is likely the cause for all your problems.

If you are unfamiliar with Git/Github, here is the direct link to the 1.5.1 libreVNA.py: https://github.com/jankae/LibreVNA/blob/v1.5.1/Documentation/UserManual/SCPI_Examples/libreVNA.py

Also, this is the programming guide for 1.5.1: https://github.com/jankae/LibreVNA/blob/v1.5.1/Documentation/UserManual/ProgrammingGuide.pdf
You will notice that certain commands (like *ESR?) are not in there yet. This also means that the 1.5.1 GUI does not support them.

**Reply by bgouveia5** [reply]
Thank you, for clearing it up for me, I wasn't understanding. From the change the program works. Thank you again for the quick response. I appreciate you!
